File: elixir/lib/bonemesh/routing.ex
defmodule Bonemesh.Routing do
  @moduledoc """
  The BoneMesh v3 routing table (protocol.md §6): direct neighbors with EWMA
  round-trip latencies (defect D3) and a distance-vector table of indirect
  destinations. Costs are real latencies; sums saturate; the node's own label is
  never a route; advertisements apply split-horizon with poisoned reverse
  (defect-D4-adjacent loop control). Pure functions over a table map.
  """

  @alpha 0.2
  @unreachable 1_000_000_000

  @doc "Sentinel path cost meaning unreachable."
  def unreachable, do: @unreachable

  @doc "A new routing table for `self_label`."
  def new(self_label), do: %{self: down(self_label), neighbors: %{}, routes: %{}}

  @doc "Records/updates a neighbor, folding in an RTT sample (EWMA)."
  def observe_neighbor(t, label, rtt) do
    key = down(label)
    ewma =
      case t.neighbors[key] do
        nil -> rtt * 1.0
        prev -> @alpha * rtt + (1.0 - @alpha) * prev
      end

    put_in(t.neighbors[key], ewma)
  end

  @doc "Removes a neighbor and withdraws routes that went through it."
  def remove_neighbor(t, label) do
    key = down(label)
    routes = t.routes |> Enum.reject(fn {_d, {via, _c}} -> via == key end) |> Map.new()
    %{t | neighbors: Map.delete(t.neighbors, key), routes: routes}
  end

  @doc "Learns/improves a route to `dest` advertised by `via` at `adv_cost`."
  def learn_route(t, dest, via, adv_cost) do
    d = down(dest)
    v = down(via)

    cond do
      d == t.self -> t
      d == v -> t
      not Map.has_key?(t.neighbors, v) -> t
      adv_cost >= @unreachable -> withdraw_if_via(t, d, v)
      true -> maybe_install(t, d, v, saturating(adv_cost, round(t.neighbors[v])))
    end
  end

  @doc "The next-hop neighbor toward `dest`, or nil."
  def next_hop(t, dest) do
    d = down(dest)

    cond do
      Map.has_key?(t.neighbors, d) -> d
      (r = t.routes[d]) != nil -> elem(r, 0)
      true -> nil
    end
  end

  @doc "The advertisement for `to_neighbor` (split-horizon + poisoned reverse)."
  def advertise_to(t, to_neighbor) do
    n = down(to_neighbor)

    neighbor_ads =
      t.neighbors
      |> Enum.reject(fn {k, _} -> k == n or k == t.self end)
      |> Map.new(fn {k, ewma} -> {k, round(ewma)} end)

    Enum.reduce(t.routes, neighbor_ads, fn {dest, {via, cost}}, acc ->
      Map.put(acc, dest, if(via == n, do: @unreachable, else: cost))
    end)
    |> Map.delete(t.self)
  end

  defp maybe_install(t, d, v, cost) do
    case t.routes[d] do
      nil -> put_in(t.routes[d], {v, cost})
      {^v, _} -> put_in(t.routes[d], {v, cost})
      {_other, old} when cost < old -> put_in(t.routes[d], {v, cost})
      _ -> t
    end
  end

  defp withdraw_if_via(t, d, v) do
    case t.routes[d] do
      {^v, _} -> %{t | routes: Map.delete(t.routes, d)}
      _ -> t
    end
  end

  defp saturating(a, b) do
    sum = a + b
    if sum >= @unreachable, do: @unreachable, else: sum
  end

  defp down(label), do: String.downcase(label)
end

File: python/bonemesh/routing.py
"""BoneMesh v3 distance-vector routing (protocol.md §5/§6).

A table of direct neighbours (EWMA-smoothed link latency) and learned routes
(destination -> next hop, path cost in ms), plus a bounded dedup set for relayed
messages. Wire-compatible with every other reference router.

Poison sentinel: every implementation now advertises exactly ``1000000000`` and
treats any advertised cost ``>= 1000000000`` as unreachable, both pinned by
protocol.md section 0. Four ports used to emit 2^63-1 and interoperated only by
luck of the tolerant threshold; this port always emitted the smaller value, which
no JSON parser can mangle.
"""

from __future__ import annotations

import math

UNREACHABLE = 1_000_000_000
POISON_THRESHOLD = 1_000_000_000
ALPHA = 0.2


class Ewma:
    def __init__(self) -> None:
        self.value = 0.0
        self.has = False

    def observe(self, sample: float) -> None:
        if not self.has:
            self.value = float(sample)
            self.has = True
        else:
            self.value = ALPHA * sample + (1 - ALPHA) * self.value

    def millis(self) -> int:
        # Half away from zero (protocol.md section 5), not Python's round(), which
        # is banker's rounding: round(2.5) is 2 here and 3 in the other six. This
        # was dismissed as a local ranking input, but the rounded value is exactly
        # what disco.routes puts on the wire, so two nodes advertised costs
        # differing by 1 ms for the same measured link. Latency is non-negative,
        # so floor(x + 0.5) is half-away-from-zero.
        return math.floor(self.value + 0.5) if self.has else UNREACHABLE


def sat_sum(a: int, b: int) -> int:
    if a >= POISON_THRESHOLD or b >= POISON_THRESHOLD:
        return UNREACHABLE
    total = a + b
    return UNREACHABLE if total >= POISON_THRESHOLD else total


class Table:
    def __init__(self, self_label: str) -> None:
        self.self_label = self_label.lower()
        self.neighbors: dict[str, Ewma] = {}
        self.routes: dict[str, dict] = {}  # dest -> {"via": label, "cost": int}

    def observe_neighbor(self, label: str, rtt_millis: float) -> None:
        k = label.lower()
        e = self.neighbors.get(k)
        if e is None:
            e = Ewma()
            self.neighbors[k] = e
        e.observe(rtt_millis)

    def remove_neighbor(self, label: str) -> None:
        k = label.lower()
        self.neighbors.pop(k, None)
        for dest in [d for d, r in self.routes.items() if r["via"] == k]:
            del self.routes[dest]

    def learn_route(self, dest: str, via: str, advertised_cost: int) -> None:
        d = dest.lower()
        v = via.lower()
        if d == self.self_label or d == v:
            return
        if v not in self.neighbors:
            return
        # Never shadow a direct neighbour with a learned route: the direct session
        # is always preferable, and letting a route win here was a real
        # convergence defect (docs/architecture.md §4).
        if d in self.neighbors:
            return
        if advertised_cost >= POISON_THRESHOLD:
            r = self.routes.get(d)
            if r and r["via"] == v:
                del self.routes[d]
            return
        cost = sat_sum(advertised_cost, self._neighbor_latency(v))
        r = self.routes.get(d)
        if r is None or r["via"] == v or cost < r["cost"]:
            self.routes[d] = {"via": v, "cost": cost}

    def next_hop(self, dest: str) -> str | None:
        d = dest.lower()
        if d in self.neighbors:
            return d  # a direct neighbour is its own next hop
        r = self.routes.get(d)
        return r["via"] if r else None

    def advertise_to(self, to_neighbor: str) -> dict:
        to = to_neighbor.lower()
        m: dict[str, int] = {}
        for label, e in self.neighbors.items():
            if label != to:
                m[label] = e.millis()
        for dest, r in self.routes.items():
            m[dest] = UNREACHABLE if r["via"] == to else r["cost"]
        m.pop(self.self_label, None)
        return m

    def route_table(self) -> dict:
        return {dest: r["via"] for dest, r in self.routes.items()}

    def _neighbor_latency(self, k: str) -> int:
        e = self.neighbors.get(k)
        return e.millis() if e else UNREACHABLE


class Dedup:
    """A bounded set of recently seen keys (message id + chunk index).

    Used to drop duplicate and looped relay traffic.
    """

    def __init__(self, cap: int) -> None:
        self.cap = cap
        self.seen: set[str] = set()
        self.order: list[str] = []

    def saw_before(self, key: str) -> bool:
        if key in self.seen:
            return True
        self.seen.add(key)
        self.order.append(key)
        if len(self.order) > self.cap:
            self.seen.discard(self.order.pop(0))
        return False
